library/std/src/sys/sync/rwlock/queue.rs

```diff
@@ -456,25 +484,83 @@ impl RwLock {
     /// * There must be threads queued on the lock.
     #[cold]
     unsafe fn unlock_contended(&self, mut state: State) {
+        debug_assert!(state.addr() & STATE == (QUEUED | LOCKED));
         loop {
             // Atomically release the lock and try to acquire the queue lock.
-            let next = state.map_addr(|a| (a & !LOCKED) | QUEUE_LOCKED);
-            match self.state.compare_exchange_weak(state, next, AcqRel, Relaxed) {
-                // The queue lock was acquired. Release it, waking up the next
-                // waiter in the process.
-                Ok(_) if state.addr() & QUEUE_LOCKED == 0 => unsafe {
-                    return self.unlock_queue(next);
-                },
+            let next = state.wrapping_byte_sub(LOCKED);
+            if state.addr() & QUEUE_LOCKED != 0 {
                 // Another thread already holds the queue lock, leave waking up
                 // waiters to it.
-                Ok(_) => return,
-                Err(new) => state = new,
+                match self.state.compare_exchange_weak(state, next, Release, Relaxed) {
+                    Ok(_) => return,
+                    Err(new) => state = new,
+                }
+            } else {
+                // Acquire the queue lock and wake up the next waiter.
+                let next = next.wrapping_byte_add(QUEUE_LOCKED);
+                match self.state.compare_exchange_weak(state, next, AcqRel, Relaxed) {
```

Comment:
Could this change to something like this?

```rust
if state.addr() & QUEUE_LOCKED != 0 {
    let next = state.wrapping_byte_sub(LOCKED);
    match self.state.compare_exchange_weak(state, next, Release, Relaxed) {
        Ok(_) => return,
        Err(new) => {
            state = new;
            continue;
        }
    }
}

let next = state.map_addr(|a| (a & !LOCKED) | QUEUE_LOCKED);
match self.state.compare_exchange_weak(state, next, AcqRel, Relaxed) {
    /* ... */
}
```

Mostly just to keep the diff smaller, but it is also a bit easier to reason with `next` this way (compared to separate `- LOCKED + QUEUE_LOCKED`).